### ISSUE TITLE
Optimistic Concurrency Control support?

### ISSUE THREAD

**Issue by rgorel** [elastic/elasticsearch-rails#347]
Hello,

I want to implement saving the documents using [Optimistic Concurrency Control](http://www.elastic.co/guide/en/elasticsearch/guide/current/optimistic-concurrency-control.html) feature of Elasticsearch, but I was not able to do it with your gem.

Considering that I have
```ruby
class Color
    include Elasticsearch::Persistence::Model
    # ...
end
```
I have tried:
```ruby
c = Search::Color.find(517053)
c.update product_strapline: 'www aaa', _version: 5
repository = Elasticsearch::Persistence::Repository.new
repository.update c, id: c.id, _version: 2
```
both with no luck, which makes me think that this feature is not supported yet.

Are there any future plans to include this feature in the next releases? Can you suggest any workaround how can I implement this using your gem?

Thanks in advance!

**Comment by karmi** [comment]
So this is related to `elasticsearch-persistence` gem, right? I don't think I have an integration test for that, which is a shame.

The options passed to the `update` method should be passed all the way down to the client call, so in theory versioning should "just work" as you outline. I'll have a look into that.

One minor note -- if you `include Elasticsearch::Persistence::Model`, there's no need to use the repository, since you already have it in the model. In other words, either you use the _repository_ pattern with lightweight model objects defined eg. with Virtus, or you use the _active_model_ pattern where everything is bundled together.

**Comment by rgorel** [comment]
Hi @karmi 

Yes, it is related to `elasticsearch-persistence`. 

> The options passed to the update method should be passed all the way down to the client call, so in theory versioning should "just work" as you outline. I'll have a look into that.

In the case of `Elasticsearch::Persistence::Model#update` it just simply does not work (i.e., update action performs without version parameter).

In the case of repository, i.e. `repository.update c, id: c.id, version: 2`, it raises the following error

```
Elasticsearch::Transport::Transport::Errors::Conflict: [409] {"error":"VersionConflictEngineException[[repository][3] [517053][_update]: version conflict, current [-1], provided [11]]","status":409}
```

regardless of the version number I am passing (even when it's correct).

**Comment by rgorel** [comment]
> One minor note -- if you include Elasticsearch::Persistence::Model, there's no need to use the repository, since you already have it in the model. In other words, either you use the repository pattern with lightweight model objects defined eg. with Virtus, or you use the active_model pattern where everything is bundled together.

I know this, I just wanted to show you how I try to do this with repository and I didn't make a separate Virtus model for this.

**Comment by karmi** [comment]
@rgorel I'm afraid I won't have much time to have a proper look at this in the next two weeks, and I've been too busy with the `elasticsearch-dsl` gem in the lats days. I'll try to have a look, but if I fail to, can you ping me in ~ two weeks?

We definitely need a good integration test for this in the gem, that's what I know for sure.

**Comment by rgorel** [comment]
@karmi ok, also I will try to contribute if I have time

**Comment by karmi** [comment]
@rgorel I have found it! It was actually a bug where in the `Rails` module, the `update` method has been redefined to allow converting dates coming in from the Rails' forms, but the `options` argument has not been passed down to the `update` method in the `Store` module.

This patch fixes that and adds unit and integration tests to cover the beahviour. I've also added RDoc examples to the `MyModel#update` method.

Many thanks for uncovering this!
